Qt swapping ctrl and cmd on OSX

There is special casing for ctl / cmd in the Qt backend on OSX

https://github.com/matplotlib/matplotlib/blob/9465538da0ebd041c3a87f7b004339c17576ae55/lib/matplotlib/backends/backend_qt.py#L49-L51

That is making the tests fail.  

Looking at a US mac keyboard on the left side I see:

```
ctrl | option | command
```

and looking at US windows keyboard I see

```
ctrl | meta (aka windows) | alt
```

however that does not tell us what keycodes they actually generate.

We used to implement this logic differently (before we unified the qt backend code) 

### 3.4.3

https://github.com/matplotlib/matplotlib/blob/919145fe9849c999aa491457c6de6faede5959c3/lib/matplotlib/backends/backend_qt5.py#L62-L67

### 3.2.2

https://github.com/matplotlib/matplotlib/blob/a1a5298b0d4a8d8230103d0fa8d369fca7fdeea0/lib/matplotlib/backends/backend_qt5.py#L70-L85

This swapping goes all the way back to 5fdd68adac02827bb28f17ea88d2a6996c4acb1d (2012).

I think we need to:

 - have a mac user with a mac keyboard veryify that the keys they hit match the keycodes we expect
 - if the keys are right to the human typing then fix the tests (probably by tweaking what the key-codes fed in are)
 - if the keys are wrong to the human typing
    - if they are consistent across mpl versions on osx: gnash our teeth in despair and figure out how we want to fix this (on one hand, we should make sure that the key-codes are consistent across platfroms, on the other hand we do not want to break existing OSX users).
    - if they changed from 3.4 to the default branch, fix the re-mapping

I also have a worry that Qt has started to fix this for us.

----

FYI: I just ran the latest tests and these are now failing for me and I'm wondering if it has to do with being on a mac? The tests look like they are swapping cmd/ctrl, so I tried swapping the modifier keys on my system and that didn't help.

```bash
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-control] - AssertionError: assert 'cmd+a' == 'ctrl+a'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-alt_control] - AssertionError: assert 'alt+cmd' == 'alt+control'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-control_alt] - AssertionError: assert 'cmd+alt' == 'ctrl+alt'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-modifier_order] - AssertionError: assert 'cmd+alt+ctrl+á' == 'ctrl+alt+super+á'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[Qt5Agg-backspace_mod] - AssertionError: assert 'cmd+backspace' == 'ctrl+backspace'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[QtAgg-control] - AssertionError: assert 'cmd+a' == 'ctrl+a'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[QtAgg-alt_control] - AssertionError: assert 'alt+cmd' == 'alt+control'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[QtAgg-control_alt] - AssertionError: assert 'cmd+alt' == 'ctrl+alt'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[QtAgg-modifier_order] - AssertionError: assert 'cmd+alt+ctrl+á' == 'ctrl+alt+super+á'
FAILED lib/matplotlib/tests/test_backend_qt.py::test_correct_key[QtAgg-backspace_mod] - AssertionError: assert 'cmd+backspace' == 'ctrl+backspace'
```

It looks like these _may_ be getting skipped on the macos-latest CI?

```bash
SKIPPED [22] lib/matplotlib/tests/test_backend_qt.py:168: No usable Qt bindings
```

_Originally posted by @greglucas in https://github.com/matplotlib/matplotlib/issues/20868#issuecomment-921432176_

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -171,37 +171,38 @@ jobs:
           # (sometimes, the install appears to be successful but shared
           # libraries cannot be loaded at runtime, so an actual import is a
           # better check).
-          if [[ "${{ runner.os }}" != 'macOS' ]]; then
-            # PyGObject, pycairo, and cariocffi do not install on OSX 10.12.
-             python -m pip install --upgrade pycairo 'cairocffi>=0.8' PyGObject &&
-               python -c 'import gi; gi.require_version("Gtk", "3.0"); from gi.repository import Gtk' &&
-               echo 'PyGObject is available' ||
-               echo 'PyGObject is not available'
+          # PyGObject, pycairo, and cariocffi do not install on OSX 10.12.
+           python -m pip install --upgrade pycairo 'cairocffi>=0.8' PyGObject &&
+             python -c 'import gi; gi.require_version("Gtk", "3.0"); from gi.repository import Gtk' &&
+             echo 'PyGObject is available' ||
+             echo 'PyGObject is not available'
 
-            # There are no functioning wheels available for OSX 10.12 (as of
-            # Sept 2020) for either pyqt5 (there are only wheels for 10.13+) or
-            # pyside2 (the latest version (5.13.2) with 10.12 wheels has a
-            # fatal to us bug, it was fixed in 5.14.0 which has 10.13 wheels)
-            python -mpip install --upgrade pyqt5${{ matrix.pyqt5-ver }} &&
-              python -c 'import PyQt5.QtCore' &&
-              echo 'PyQt5 is available' ||
-              echo 'PyQt5 is not available'
+          # There are no functioning wheels available for OSX 10.12 (as of
+          # Sept 2020) for either pyqt5 (there are only wheels for 10.13+) or
+          # pyside2 (the latest version (5.13.2) with 10.12 wheels has a
+          # fatal to us bug, it was fixed in 5.14.0 which has 10.13 wheels)
+          python -mpip install --upgrade pyqt5${{ matrix.pyqt5-ver }} &&
+            python -c 'import PyQt5.QtCore' &&
+            echo 'PyQt5 is available' ||
+            echo 'PyQt5 is not available'
+          if [[ "${{ runner.os }}" != 'macOS' ]]; then
             python -mpip install --upgrade pyside2 &&
               python -c 'import PySide2.QtCore' &&
               echo 'PySide2 is available' ||
               echo 'PySide2 is not available'
-            # Qt6 crashes on Github's ubuntu 18.04 runner.
-            if [[ "${{ matrix.os }}" = ubuntu-20.04 ]]; then
-              python -mpip install --upgrade pyqt6 &&
-                python -c 'import PyQt6.QtCore' &&
-                echo 'PyQt6 is available' ||
-                echo 'PyQt6 is not available'
-              python -mpip install --upgrade pyside6 &&
-                python -c 'import PySide6.QtCore' &&
-                echo 'PySide6 is available' ||
-                echo 'PySide6 is not available'
-            fi
           fi
+          # Qt6 crashes on Github's ubuntu 18.04 runner.
+          if [[ "${{ matrix.os }}" = ubuntu-20.04 ]]; then
+            python -mpip install --upgrade pyqt6 &&
+              python -c 'import PyQt6.QtCore' &&
+              echo 'PyQt6 is available' ||
+              echo 'PyQt6 is not available'
+            python -mpip install --upgrade pyside6 &&
+              python -c 'import PySide6.QtCore' &&
+              echo 'PySide6 is available' ||
+              echo 'PySide6 is not available'
+          fi
+
           python -mpip install --upgrade \
             -f "https://extras.wxpython.org/wxPython4/extras/linux/gtk3/${{ matrix.os }}" \
             wxPython &&

--- a/lib/matplotlib/tests/test_backend_qt.py
+++ b/lib/matplotlib/tests/test_backend_qt.py
@@ -166,24 +166,33 @@ def test_fig_sigint_override(qt_core):
 
 
 @pytest.mark.parametrize(
-    'qt_key, qt_mods, answer',
+    "qt_key, qt_mods, answer",
     [
-        ('Key_A', ['ShiftModifier'], 'A'),
-        ('Key_A', [], 'a'),
-        ('Key_A', ['ControlModifier'], 'ctrl+a'),
-        ('Key_Aacute', ['ShiftModifier'],
-         '\N{LATIN CAPITAL LETTER A WITH ACUTE}'),
-        ('Key_Aacute', [],
-         '\N{LATIN SMALL LETTER A WITH ACUTE}'),
-        ('Key_Control', ['AltModifier'], 'alt+control'),
-        ('Key_Alt', ['ControlModifier'], 'ctrl+alt'),
-        ('Key_Aacute', ['ControlModifier', 'AltModifier', 'MetaModifier'],
-         'ctrl+alt+meta+\N{LATIN SMALL LETTER A WITH ACUTE}'),
+        ("Key_A", ["ShiftModifier"], "A"),
+        ("Key_A", [], "a"),
+        ("Key_A", ["ControlModifier"], ("ctrl+a")),
+        (
+            "Key_Aacute",
+            ["ShiftModifier"],
+            "\N{LATIN CAPITAL LETTER A WITH ACUTE}",
+        ),
+        ("Key_Aacute", [], "\N{LATIN SMALL LETTER A WITH ACUTE}"),
+        ("Key_Control", ["AltModifier"], ("alt+control")),
+        ("Key_Alt", ["ControlModifier"], "ctrl+alt"),
+        (
+            "Key_Aacute",
+            ["ControlModifier", "AltModifier", "MetaModifier"],
+            ("ctrl+alt+meta+\N{LATIN SMALL LETTER A WITH ACUTE}"),
+        ),
         # We do not currently map the media keys, this may change in the
         # future.  This means the callback will never fire
-        ('Key_Play', [], None),
-        ('Key_Backspace', [], 'backspace'),
-        ('Key_Backspace', ['ControlModifier'], 'ctrl+backspace'),
+        ("Key_Play", [], None),
+        ("Key_Backspace", [], "backspace"),
+        (
+            "Key_Backspace",
+            ["ControlModifier"],
+            "ctrl+backspace",
+        ),
     ],
     ids=[
         'shift',
@@ -216,6 +225,11 @@ def test_correct_key(backend, qt_core, qt_key, qt_mods, answer):
     Assert sent and caught keys are the same.
     """
     from matplotlib.backends.qt_compat import _enum, _to_int
+
+    if sys.platform == "darwin" and answer is not None:
+        answer = answer.replace("ctrl", "cmd")
+        answer = answer.replace("control", "cmd")
+        answer = answer.replace("meta", "ctrl")
     result = None
     qt_mod = _enum("QtCore.Qt.KeyboardModifier").NoModifier
     for mod in qt_mods:

--- a/lib/matplotlib/tests/test_backend_qt.py
+++ b/lib/matplotlib/tests/test_backend_qt.py
@@ -93,10 +93,10 @@ def test_other_signal_before_sigint(qt_core, platform_simulate_ctrl_c,
     def custom_sigpipe_handler(signum, frame):
         nonlocal sigcld_caught
         sigcld_caught = True
-    signal.signal(signal.SIGCLD, custom_sigpipe_handler)
+    signal.signal(signal.SIGCHLD, custom_sigpipe_handler)
 
     def fire_other_signal():
-        os.kill(os.getpid(), signal.SIGCLD)
+        os.kill(os.getpid(), signal.SIGCHLD)
 
     def fire_sigint():
         platform_simulate_ctrl_c()

